### PR TITLE
Adds auxilary functions needed for reduce_sum

### PR DESCRIPTION
**Pull request by SteveBronder**
## Summary

Adds the following functions needed for `reduce_sum`

1. `accumulate_adjoints`: Iterates through a parameter pack and assigns any var object or containers of vars adjoint values into a contiguous block of memory.
2. `count_vars`:  Iterates over a parameter pack counting the number of var objects as well as vars in containers
3. `deep_copy_vars`: Iterates over a parameter pack while only copying the var objects and var containers.
4. `save_varis` Iterates over a paramater pack of var objects and containers storing the vari pointers into a contiguous block of memory

## Tests

Tests can be run with 

```bash
./runTests.py test/unit/math/rev/core/accumulate_adjoints_test.cpp \
test/unit/math/rev/core/count_vars_test.cpp \
test/unit/math/rev/core/deep_copy_vars_test.cpp \
test/unit/math/rev/core/save_varis_test.cpp
```

## Side Effects

None

## Checklist

- [x] Math issue #1616

- [x] Copyright holder: Steve Bronder, Ben Bales, Sebastian Weber

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by bbbales2** [comment]
Ooops, missed this. Thanks for the ping! I'll get on it tonight!

**Comment by bbbales2** [comment]
Oh wait I just wrote this code and the tests lol. @wds15 can you do the review?

**Comment by wds15** [comment]
Ok...I should be able to review.

**Comment by bbbales2** [comment]
@wds15 I should probably be the one to finish up the reduce_sum tests, Steve can manage the pull, and you can review?

I'll try to get that done before the Stan meeting tomorrow. I didn't see any sign of a Math meeting.

**Comment by SteveBronder** [comment]
@bbbales2 are you handling the tests here or should I go at the above?

File: stan/math/rev/core/save_varis.hpp [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+#include <stan/math/rev/meta.hpp>
+#include <stan/math/rev/core/var.hpp>
+
+#include <utility>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+template <typename... Pargs>
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/rev/core/save_varis.hpp]: We could include a comment on the last one with a link to an [article](https://raymii.org/s/snippets/Cpp_variadic_template_recursive_example.html) explaining variadic parameter packs, how do you feel about that? We could also have a module for `reduce_sum` so on the site there's a section that explains the more complex parts of the reduce sum implimentation

[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/rev/core/save_varis.hpp]: By module I mean a section like we have for the requires stuff

http://mc-stan.org/math/d1/db9/group__require__meta.html

**Comment by bbbales2** [comment]
@SteveBronder I have a plan for deep copy.

The other things have at em'. There should be save_varis test in the reduce_sum WIP pull.

File: test/unit/math/rev/core/accumulate_adjoints_test.cpp [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+      arg1, arg18, arg17, arg2, arg16, arg3, arg15, arg4, arg14, arg5, arg13,
+      arg12, arg6, arg11, arg7, arg10, arg8, arg9);
+
+  for (int i = 0; i < num_vars; ++i)
+    EXPECT_FLOAT_EQ(storage(i), 1.0);
+  for (int i = num_vars; i < storage.size(); ++i)
+    EXPECT_FLOAT_EQ(storage(i), 0.0);
+
+  EXPECT_EQ(ptr, storage.data() + num_vars);
+}
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply test/unit/math/rev/core/accumulate_adjoints_test.cpp]: Do you just mean checking the end of the recursion method like

```
TEST(AgradRev_accumulate_adjoints, zero_args) {
  Eigen::VectorXd storage = Eigen::VectorXd::Zero(1000);
  double* ptr = stan::math::accumulate_adjoints(storage.data());

  for (int i = 0; i < storage.size(); ++i) {
    EXPECT_FLOAT_EQ(storage(i), 0.0);
  }

  EXPECT_EQ(ptr, storage.data());
}
```

**Comment by SteveBronder** [comment]
@bbbales2 do you think it makes sense to do `accumulate_adjoints_impl` like we do for `count_varis` then just have one top level function `accumulate_adjoints` exposed?

File: test/unit/math/rev/core/accumulate_adjoints_test.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (bbbales2) [reply test/unit/math/rev/core/accumulate_adjoints_test.cpp]: That looks good to me. We want to make sure it returns the pointer it got, and at least hope it didn't do anything to the data there.

[the remark this answers is not in the document]
Comment (wds15) [reply test/unit/math/rev/core/accumulate_adjoints_test.cpp]: Correct.

File: test/unit/math/rev/core/accumulate_adjoints_test.cpp [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+  var arg(5.0);
+  arg.vi_->adj_ = 1.0;
+
+  storage.setZero();
+  double* ptr = stan::math::accumulate_adjoints(storage.data(), arg);
+
+  size_t num_vars = stan::math::count_vars(arg);
+
+  for (int i = 0; i < num_vars; ++i)
+    EXPECT_FLOAT_EQ(storage(i), i + 1.0);
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply test/unit/math/rev/core/accumulate_adjoints_test.cpp]: Yep! in this loop  `i` only ever equals zero so I fixed that up here and in the other tests where that happens

**Comment by bbbales2** [comment]
> do you think it makes sense to do accumulate_adjoints_impl

I guess I have no opinion really. Maybe it makes docs clearer cause then there's only one entry point? Either way is good.

**Comment by bbbales2** [comment]
I made the changes to the deep_copy_vars tests. Hopefully they're good now.

**Comment by SteveBronder** [comment]
@wds15 this is ready for review!

File: stan/math/rev/core/deep_copy_vars.hpp [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+/**
+ * Forward arguments that do not contain vars. There
+ *   is no copying to be done.
+ *
+ * @tparam Arith an arithmetic type.
+ * @param arg For lvalue references this will be passed by reference.
+ *  Otherwise it will be moved.
+ */
+template <typename Arith, typename = require_arithmetic_t<scalar_type_t<Arith>>>
+inline decltype(auto) deep_copy_vars(Arith&& arg) {
```
Comment (wds15) [inline stan/math/rev/core/deep_copy_vars.hpp]: What do we get if the tuple passed into the function is an empty one? Does this function then qualify? If yes, then can we please test it - I think we don't do that atm.

File: stan/math/rev/core/save_varis.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (wds15) [reply stan/math/rev/core/save_varis.hpp]: Sounds great to me to have such a module.

File: test/unit/math/rev/core/save_varis_test.cpp [hunk 5]
```diff
@@ -0,0 +1,414 @@
+#include <stan/math.hpp>
+#include <gtest/gtest.h>
+#include <vector>
+
+using stan::math::var;
+using stan::math::vari;
+
+TEST(AgradRev_save_varis, int_arg) {
```
Comment (wds15) [inline test/unit/math/rev/core/save_varis_test.cpp]: I think the recursion ending function is not tested for save_varis. Can u add that please?

File: stan/math/rev/core/count_vars.hpp [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+#include <utility>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+template <typename VecVar, require_std_vector_vt<is_var, VecVar>* = nullptr,
+          typename... Pargs>
+inline size_t count_vars_impl(size_t count, VecVar&& x, Pargs&&... args);
```
Comment (wds15) [inline stan/math/rev/core/count_vars.hpp]: Sorry for bringing this up somewhat late.... but shouldn't we place the "*_impl" stuff into an "internal" namespace? This applies to the other tools as well which we are writing with an impl naming.

Comment (wds15) [reply stan/math/rev/core/count_vars.hpp]: Oh... count_vars is the only one with a "impl" thing - for good reasons as here you have to initialise to zero. So, I think moving the "*_impl" functions into an internal namespace, makes sense to me. Right?

**Comment by bbbales2** [comment]
Gonna make these changes real quick

File: stan/math/rev/core/deep_copy_vars.hpp (hunk 4 above)
Comment (bbbales2) [reply stan/math/rev/core/deep_copy_vars.hpp]: I added an empty tuple test to the apply function.

I think this is maybe a higher level question that deep copy. I think it's covered by this reduce_sum test (what happens when there are no local args): https://github.com/stan-dev/math/blob/6743ff6937159857b1409f8ee15c7c9dc7589670/test/unit/math/mix/functor/reduce_sum_test.cpp#L63

**Comment by bbbales2** [comment]
@wds15 Ready to check again (if tests pass)

Also what was "Sounds great to me to have such a module." pointing at?

**Comment by wds15** [comment]
Steve suggested to add some documentation on the tuples argument processing magic as documentation to the modules doxygen section.

**Comment by wds15** [comment]
Looks like we have issue with `pow`??? Stan unit tests are failing with:

```
In file included from src/test/unit/services/sample/hmc_static_unit_e_adapt_test.cpp:4:
./test/test-models/good/optimization/rosenbrock.hpp:177:31: error: call to 'pow' is ambiguous
            lp_accum__.add(-((pow((1 - x), 2) + (100 * pow((y - pow(x, 2)), 2)))));
                              ^~~
src/stan/services/util/initialize.hpp:127:33: note: in instantiation of function template specialization 'rosenbrock_model_namespace::rosenbrock_model::log_prob<false, true, double>' requested here
      log_prob = model.template log_prob<false, Jacobian>(unconstrained,
                                ^
src/stan/services/sample/hmc_static_unit_e_adapt.hpp:62:43: note: in instantiation of function template specialization 'stan::services::util::initialize<true, rosenbrock_model_namespace::rosenbrock_model, boost::random::additive_combine_engine<boost::random::linear_congruential_engine<unsigned int, 40014, 0, 2147483563>, boost::random::linear_congruential_engine<unsigned int, 40692, 0, 2147483399> > >' requested here
  std::vector<double> cont_vector = util::initialize(
                                          ^
```

**Comment by wds15** [comment]
Same issue here: https://github.com/stan-dev/math/pull/1811

Maybe something odd was merged to develop or there are some issues with the Jenkins pipelines... no idea. @rok-cesnovar do you have a hint? Not sure who else to ping.

**Comment by rok-cesnovar** [comment]
Yes, this is the same issue that happened in #1804. Try restarting tests and hope the Linux machine does not get picked.

See discussion https://github.com/stan-dev/math/pull/1804#issuecomment-605824397

Ben is on it, I was hoping #1810 would fix it, but it apparently needs some more work. My opinion is that we give Ben some more time to fix it, otherwise we revert.

**Comment by wds15** [comment]
Thanks. I wasn't aware of this mess. Looks ugly, but let's see. Is the complex stuff what is causing the hiccup here?

**Comment by rok-cesnovar** [comment]
Yes, its from the complex PR. It fails on a specific Linux AWS EC2 machine. If that test runs elsewhere it's fine. Its an issue with overloads and standard libraries I believe.

**Pull request merged**